File: test/unicode_tokenization_test.dart
import 'package:flutter_test/flutter_test.dart';
import '../lib/core/normalize.dart';

void main() {
  group('Unicode Tokenization Tests', () {
    test('should tokenize Spanish text with diacritics', () {
      final tokens = tokenizeUnicode('canción');
      expect(tokens, equals(['canción']));
    });

    test('should tokenize German text with umlauts', () {
      final tokens = tokenizeUnicode('Über');
      expect(tokens, equals(['Über']));
    });

    test('should tokenize Hebrew text', () {
      final tokens = tokenizeUnicode('שלום עולם');
      expect(tokens, equals(['שלום', 'עולם']));
    });

    test('should tokenize mixed script text', () {
      final tokens = tokenizeUnicode('hello123世界');
      expect(tokens, equals(['hello123世界']));
    });

    test('should handle text with digits inline', () {
      final tokens = tokenizeUnicode('test123word');
      expect(tokens, equals(['test123word']));
    });

    test('should maintain legacy ASCII tokenization behavior', () {
      final tokens = tokenizeUnicode('bad word');
      expect(tokens, equals(['bad', 'word']));
    });

    test('should handle empty and whitespace-only text', () {
      expect(tokenizeUnicode(''), equals([]));
      expect(tokenizeUnicode('   '), equals([]));
      expect(tokenizeUnicode('\t\n'), equals([]));
    });

    test('should handle punctuation boundaries correctly', () {
      final tokens = tokenizeUnicode('hello, world! test.');
      expect(tokens, equals(['hello', 'world', 'test']));
    });

    test('should preserve Unicode combining marks', () {
      // Test with combining diacritical marks
      final tokens = tokenizeUnicode('café');
      expect(tokens.isNotEmpty, true);
      expect(tokens.first.contains('é'), true);
    });

    test('should handle Arabic text', () {
      final tokens = tokenizeUnicode('مرحبا بالعالم');
      expect(tokens, equals(['مرحبا', 'بالعالم']));
    });
  });

  group('Legacy normalizeWord tests', () {
    test('should still normalize accented characters', () {
      expect(normalizeWord('café'), equals('cafe'));
      expect(normalizeWord('naïve'), equals('naive'));
      expect(normalizeWord('piñata'), equals('pinata'));
    });
  });
}

File: test/UnicodeTokenizationTest.kt
import org.junit.Test
import org.junit.Assert.*

/**
 * Simple test to validate Unicode tokenization in Android Kotlin
 * Run with: kotlinc-jvm UnicodeTokenizationTest.kt -cp .:junit-4.12.jar && kotlin UnicodeTokenizationTestKt
 */
class UnicodeTokenizationTest {
    
    private fun tokenizeUnicode(text: String): List<String> {
        // Split on non-letter, non-mark, non-digit boundaries (Unicode-aware)
        return text.split(Regex("[^\\p{L}\\p{M}\\p{N}]+")).filter { it.isNotBlank() }
    }

    @Test
    fun testSpanishWithDiacritics() {
        val tokens = tokenizeUnicode("canción")
        assertEquals(listOf("canción"), tokens)
    }

    @Test
    fun testGermanUmlauts() {
        val tokens = tokenizeUnicode("Über")
        assertEquals(listOf("Über"), tokens)
    }

    @Test
    fun testHebrewText() {
        val tokens = tokenizeUnicode("שלום עולם")
        assertEquals(listOf("שלום", "עולם"), tokens)
    }

    @Test
    fun testMixedScript() {
        val tokens = tokenizeUnicode("hello123世界")
        assertEquals(listOf("hello123世界"), tokens)
    }

    @Test
    fun testDigitsInline() {
        val tokens = tokenizeUnicode("test123word")
        assertEquals(listOf("test123word"), tokens)
    }

    @Test
    fun testLegacyBehavior() {
        val tokens = tokenizeUnicode("bad word")
        assertEquals(listOf("bad", "word"), tokens)
    }

    @Test
    fun testEmptyAndWhitespace() {
        assertEquals(emptyList<String>(), tokenizeUnicode(""))
        assertEquals(emptyList<String>(), tokenizeUnicode("   "))
        assertEquals(emptyList<String>(), tokenizeUnicode("\t\n"))
    }

    @Test
    fun testPunctuationBoundaries() {
        val tokens = tokenizeUnicode("hello, world! test.")
        assertEquals(listOf("hello", "world", "test"), tokens)
    }

    @Test
    fun testArabicText() {
        val tokens = tokenizeUnicode("مرحبا بالعالم")
        assertEquals(listOf("مرحبا", "بالعالم"), tokens)
    }
}